### PR TITLE
Convert constructor types in py structures

### PR DESCRIPTION
**Pull request by lucacarniato**
…linearGrid constructor parameters to appropriate types in case the user does not specify the correct types

When the user types are not convertible, asarray throws a ValueError exception

### CODE REVIEW

File: tests/test_py_structures_inputs.py [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+    assert_array_equal(mesh_2d.node_y, np.array([4.0, 5.0, 6.0], dtype=np.double))
+    assert_array_equal(mesh_2d.edge_nodes, np.array([0, 1, 1, 2], dtype=np.int32))
+    assert_array_equal(mesh_2d.face_nodes, np.array([0, 1, 2], dtype=np.int32))
+    assert_array_equal(mesh_2d.nodes_per_face, np.array([3, 4, 5], dtype=np.int32))
+    assert_array_equal(mesh_2d.edge_x, np.array([1.5, 2.5], dtype=np.double))
+    assert_array_equal(mesh_2d.edge_y, np.array([4.5, 5.5], dtype=np.double))
+    assert_array_equal(mesh_2d.face_x, np.array([2.0, 3.0], dtype=np.double))
+    assert_array_equal(mesh_2d.face_y, np.array([5.0, 6.0], dtype=np.double))
+    assert_array_equal(mesh_2d.edge_faces, np.array([0, 1], dtype=np.int32))
+    assert_array_equal(mesh_2d.face_edges, np.array([2, 3], dtype=np.int32))
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:86]: Optional: The expected array can be `np.asarray(input, dtype=np.double)` rather explicitly creating an array with the expected dtype. `input` is one of the vars defined in lines 50-60.

File: tests/test_py_structures_inputs.py [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+    assert splines_to_curvilinear_parameters.aspect_ratio_grow_factor == 1.1
+    assert splines_to_curvilinear_parameters.average_width == 500.0
+    assert splines_to_curvilinear_parameters.curvature_adapted_grid_spacing == 1
+    assert splines_to_curvilinear_parameters.grow_grid_outside == 0
+    assert splines_to_curvilinear_parameters.maximum_num_faces_in_uniform_part == 5
+    assert (
+        splines_to_curvilinear_parameters.nodes_on_top_of_each_other_tolerance == 0.0001
+    )
+    assert splines_to_curvilinear_parameters.min_cosine_crossing_angles == 0.95
+    assert splines_to_curvilinear_parameters.check_front_collisions == 0
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:307]: bool, False

File: tests/test_py_structures_inputs.py [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+    assert splines_to_curvilinear_parameters.remove_skinny_triangles == 1
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:308]: bool, True

File: tests/test_py_structures_inputs.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+        min_cosine_crossing_angles="0.95",
+        check_front_collisions="0",
+        remove_skinny_triangles="1",
+    )
+
+    assert splines_to_curvilinear_parameters.aspect_ratio == 0.1
+    assert splines_to_curvilinear_parameters.aspect_ratio_grow_factor == 1.1
+    assert splines_to_curvilinear_parameters.average_width == 500.0
+    assert splines_to_curvilinear_parameters.curvature_adapted_grid_spacing == 1
+    assert splines_to_curvilinear_parameters.grow_grid_outside == 0
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:301]: bool, False

File: tests/test_py_structures_inputs.py [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+
+    node_x = ["1.0", "2.0", "3.0"]
+    node_y = ["4.0", "5.0", "6.0"]
+    edge_nodes = ["0", "1", "1", "2"]
+
+    mesh_1d = Mesh1d(node_x=node_x, node_y=node_y, edge_nodes=edge_nodes)
+
+    assert np.array_equal(mesh_1d.node_x, np.array([1.0, 2.0, 3.0], dtype=np.double))
+    assert np.array_equal(mesh_1d.node_y, np.array([4.0, 5.0, 6.0], dtype=np.double))
+    assert np.array_equal(mesh_1d.edge_nodes, np.array([0, 1, 1, 2], dtype=np.int32))
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:420]: Optional: can use numpy.asarray

File: tests/test_py_structures_inputs.py [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+    """Test implicit conversion from string to double for Contacts works"""
+
+    # Test valid input
+    mesh1d_indices = ["0", "1", "2"]
+    mesh2d_indices = ["0", "1", "2"]
+
+    contacts = Contacts(mesh1d_indices=mesh1d_indices, mesh2d_indices=mesh2d_indices)
+
+    assert np.array_equal(contacts.mesh1d_indices, np.array([0, 1, 2], dtype=np.int32))
+    assert np.array_equal(contacts.mesh2d_indices, np.array([0, 1, 2], dtype=np.int32))
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:448]: Optional: can use numpy.asarray

File: tests/test_py_structures_inputs.py [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+
+
+def test_contacts_invalid_input():
+    """Test exceptions due to malformed input for Contacts"""
+    with pytest.raises(ValueError):
+        Contacts(mesh1d_indices=["0", "1", "abc"], mesh2d_indices=["0", "1", "2"])
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:454]: Please replace the silly string "abc" with something more meaningful, elegant and powerful. The string "nonna" is recommended.

File: tests/test_py_structures_inputs.py [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+    )
+    assert np.array_equal(
+        gridded_samples.y_coordinates, np.array([2.0, 3.0, 4.0], dtype=np.double)
+    )
+    assert np.array_equal(
+        gridded_samples.values,
+        np.array(
+            [10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0],
+            dtype=np.double,
+        ),
```
Comment (ahmad-el-sayed) [inline tests/test_py_structures_inputs.py:512]: Optional: can use numpy.asarray

File: tests/test_py_structures_inputs.py [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+
+    output_mesh2d = mk.mesh2d_get()
+
+    # Test if the input and output differs
+    assert_array_equal(output_mesh2d.edge_nodes, input_mesh2d.edge_nodes)
+    assert_array_equal(output_mesh2d.node_x, input_mesh2d.node_x)
+    assert_array_equal(output_mesh2d.node_y, input_mesh2d.node_y)
+
+    # Test if faces are correctly calculated
+    assert_array_equal(output_mesh2d.face_nodes, np.array([0, 1, 2, 3]))
```
[the remark this answers is not in the document]
Comment (lucacarniato) [reply tests/test_py_structures_inputs.py:42]: sorry i forgot, in the lines below i have not

File: tests/test_py_structures_inputs.py (hunk 7 above)
Comment (lucacarniato) [reply tests/test_py_structures_inputs.py:454]: ahaha

**Review by ahmad-el-sayed — APPROVED** [review]

**Pull request merged**